File: scripts/node/parsers/genAiPromptTemplateParser.js
'use strict';

const fs = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads GenAI Prompt Template (Prompt Builder / Agentforce) metadata.
 * @param {string} typeDir - path to the genAiPromptTemplates/ folder
 * @returns {{ templates: Array }}
 */
async function parseGenAiPromptTemplates(typeDir) {
  const templates = [];
  if (!fs.existsSync(typeDir)) return { templates };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.genAiPromptTemplate-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    templates.push({
      apiName: file.replace('.genAiPromptTemplate-meta.xml', ''),
      masterLabel: extractTag(xml, 'masterLabel'),
      templateType: extractTag(xml, 'templateType'),
      description: extractTag(xml, 'description'),
      content: extractTag(xml, 'content'),
    });
  }

  return { templates };
}

module.exports = { parseGenAiPromptTemplates };

File: scripts/python/parsers/gen_ai_prompt_template_parser.py
"""
gen_ai_prompt_template_parser.py
Reads Salesforce GenAI Prompt Template (Prompt Builder) metadata from SFDX source format.
Parses .genAiPromptTemplate-meta.xml files.
"""

import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_gen_ai_prompt_templates(type_dir: str) -> dict:
    """Parse all *.genAiPromptTemplate-meta.xml files in a genAiPromptTemplates/ folder."""
    templates = []
    root = Path(type_dir)
    if not root.exists():
        return {'templates': templates}

    for file in sorted(root.glob('*.genAiPromptTemplate-meta.xml')):
        xml = file.read_text(encoding='utf-8')
        templates.append({
            'apiName': file.name.replace('.genAiPromptTemplate-meta.xml', ''),
            'masterLabel': _extract_tag(xml, 'masterLabel'),
            'templateType': _extract_tag(xml, 'templateType'),
            'description': _extract_tag(xml, 'description'),
            'content': _extract_tag(xml, 'content'),
        })

    return {'templates': templates}
